service/version.go

```diff
@@ -0,0 +1,55 @@
+package service
+
+import (
+	"fmt"
+	"strings"
+
+	"golang.org/x/mod/semver"
+
+	"github.com/canonical/microcloud/microcloud/api/types"
+)
+
+const (
+	// lxdMinVersion is the minimum version of LXD that fully supports all MicroCloud features.
+	lxdMinVersion = "6.1"
+
+	// microCephMinVersion is the minimum version of MicroCeph that fully supports all MicroCloud features.
+	microCephMinVersion = "19.2"
+
+	// microOVNMinVersion is the minimum version of MicroOVN that fully supports all MicroCloud features.
+	microOVNMinVersion = "24.03"
+)
+
+// validateVersion checks that the daemon version for the given service is at a supported version for this version of MicroCloud.
+func validateVersion(serviceType types.ServiceType, daemonVersion string) error {
+	switch serviceType {
+	case types.LXD:
+		lxdVersion := semver.Canonical(fmt.Sprintf("v%s", daemonVersion))
+		expectedVersion := semver.Canonical(fmt.Sprintf("v%s", lxdMinVersion))
+		if semver.Compare(semver.MajorMinor(lxdVersion), semver.MajorMinor(expectedVersion)) != 0 {
+			return fmt.Errorf("%s version %q is not supported", serviceType, daemonVersion)
+		}
+
+	case types.MicroOVN:
+		if daemonVersion != microOVNMinVersion {
+			return fmt.Errorf("%s version %q is not supported", serviceType, daemonVersion)
+		}
+
+	case types.MicroCeph:
+		// Parse out the underlying ceph version from the version string.
+		before, after, ok := strings.Cut(daemonVersion, "ceph-version: ")
+		cephVersion := strings.Split(after, "~")
+		if !ok || before != "" || !strings.Contains(after, "~") || len(cephVersion) != 2 || cephVersion[0] == "" {
```

Comment:
I was thinking a bit on this one whether or not it would make sense to use a regex instead to only match the version at the beginning of the string so we are protected against upstream version schema changes and don't have to deal with the `~` and splitting on it.

But if that happens we can possibly react on it quite quickly anyway so not a blocker.

Reply:
If there's a schema change, I wouldn't think it's safe to assume the first encountered semver will correspond to what we care about, we'd still have to verify manually. So in that sense I think it's fine to do string manipulation since it's easier to see what is happening to the string.